### PR TITLE
Preparation for repo sharding

### PR DESCRIPTION
**Pull request by segiddins**
~~The migration script is still unfinished, so please do not merge yet~~

This should be ready to go!

### CODE REVIEW

File: script/shard_spec_repo [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+Pod::Executable.execute_command(:heroku, %w(config:set TRUNK_APP_PUSH_ALLOWED=false), true)
+
+Pod::Command.run(%w(repo update master))
+Pod::Command.run(%W(repo shard master --lengths=#{prefix_lengths.join(',')}))
+
+shard_sha = Dir.chdir(master.repo) { Pod::Executable.execute_command(:git, %w(rev-parse HEAD), true).strip }
+
+# creating from scratch to avoid re-serializing any stale data
+Pod::Source::Metadata.from_file(master.metadata_path).tap do |metadata|
+  metadata.minimum_cocoapods_version = Pod::Version.new('1.0.0.beta.6')
```
Comment (orta) [inline script/shard_spec_repo]: _probably_ better to have that as a non-beta? I know that's when we added support, but probably safer

File: script/shard_spec_repo [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  master.metadata_path.open('w') { |f| f.write(yaml) }
+  Pod::Executable.execute_command(:heroku, %W(config:set MASTER_SOURCE_METADATA=#{yaml}), true)
+end
+
+Dir.chdir(master.repo) do
+  Pod::Executable.execute_command(:git, %W(add #{master.metadata_path}), true)
+  Pod::Executable.execute_command(:git, ['commit',
+                                         '-m', 'Update the minimum cocoapods version to 1.0',
+                                         '-m', 'See http://blog.cocoapods.org/Master-Spec-Repo-Rate-Limiting-Post-Mortem/ for more information'],
+                                  true)
```
Comment (orta) [inline script/shard_spec_repo]: nice trick, new to me

File: script/shard_spec_repo (hunk 1 above)
Comment (segiddins) [inline script/shard_spec_repo]: it's just the minimum, in case someone is stuck on that bates for some reason?

**Comment by orta** [comment]
The crux of the commit choice is tough, _on trunk_ we only do one thing with the commit reference - we pull the JSON for a Pod out of it,

> Spec > Spec Version > Commit > JSON attributes

I don't think we need to update all the commits

**Comment by orta** [comment]
We have orders [discussion in slack too](https://cocoapods.slack.com/archives/web/p1473781279000052)

File: script/shard_spec_repo (hunk 1 above)
Comment (orta) [inline script/shard_spec_repo]: My only worry is that older CocoaPods versions have had issues comparing beta SemVer versions

**Comment by segiddins** [comment]
> IMO a single commit should be fine.
What do you mean by that?

**Comment by orta** [comment]
it was re:

```
+# Two options here, and I'm not sure which is best:
 +# 1) Update the `commits.sha` for the latest commit of each `pod_version`
 +#    to be `shard_sha`, which destroys the fact that the `sha` generally points
 +#    to the exact commit in which the version was added/updated
 +# 2) Create a new `commit` row for all 100,000+ `pod_version`s.
 +#    This is semantically more correct, but since each commit stores the spec
 +#    JSON, this has the potential to significantly bloat the DB

```

File: script/shard_spec_repo (hunk 1 above)
Comment (segiddins) [inline script/shard_spec_repo]: we fixed that bug, and anything pre-1.0 will fail in the first, numeric, comparison

**Comment by segiddins** [comment]
Right, but I think we then came to the conclusion that we wouldn't need to add _any_ new commit rows?

**Comment by orta** [comment]
Yes, no new rows in the db

**Comment by segiddins** [comment]
> Yes, no new rows in the db - so the difference is mainly in the .git's semantics

I'm not sure what you mean by this

**Comment by orta** [comment]
I don't think either of the options matter from my perspective, as there'll be no changes to the trunk db.

**Comment by segiddins** [comment]
This is ready for finally review, @orta @alloy @DanToml

File: app/models/pod_info.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      end
+
+      def self.versions
+        pods = Hash.new {|h,k| h[k] = CompactIndex::Gem.new(k, []) }
+        PodVersion.each do |v|
+          ::Pod::Specification.from_json(v.last_published_by.specification_data).recursive_subspecs.each do |ss|
+            pods[ss.name].versions << CompactIndex::GemVersion.new(v.name, nil, nil, 'info_cs')
+          end
+        end
+        CompactIndex.versions(CompactIndex::VersionsFile.new("/Users/segiddins/Development/Strata/trunk.cocoapods.org/versions.list"), pods.values)
```
Comment (orta) [inline app/models/pod_info.rb]: should this be relative?

File: script/shard_spec_repo [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      FileUtils.cp_r %w(Specs .gitignore README.md CocoaPods-version.yml), '.'
+      Pod::Executable.execute_command(:git, %w(init .), true)
+      Pod::Executable.execute_command(:git, %w(remote add old-specs https://github.com/CocoaPods/Old-Specs.git), true)
+      Pod::Executable.execute_command(:git, %w(add -A), true)
+      Pod::Executable.execute_command(:git, ['commit',
+                                             '-m', "Create a copy of CocoaPods/Specs@#{pre_shard_sha}",
+                                             '-m', 'Please see https://blog.cocoapods.org/Sharding/ for more information',
+                                             '-m', "Please use https://github.com/CocoaPods/Specs if you're running CocoaPods 1.0+"],
+                                      true)
+      Pod::Executable.execute_command(:git, %w(push old-specs master), true)
```
Comment (orta) [inline script/shard_spec_repo]: 👍 this is awesome

File: script/shard_spec_repo [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                                         '-m', 'See http://blog.cocoapods.org/Master-Spec-Repo-Rate-Limiting-Post-Mortem/ for more information',
+                                         '-m', 'See http://blog.cocoapods.org/Sharding/ for the announcement of sharding',
+                                         '-m', "If you're stuck on CocoaPods < 1.0, use the `https://github.com/CocoaPods/Old-Specs.git` source from now on"],
+                                  true)
+  Pod::Executable.execute_command(:git, %w(push origin master), true)
+end
+
+# We purposefully don't create new commit objects, as the old commits will still
+# be available on the repo and we'd like to avoid severely bloating the
+# database.
```
Comment (orta) [inline script/shard_spec_repo]: 👍

File: app/models/pod_info.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
+require 'compact_index'
+
+module Pod
+  module TrunkApp
+    module PodInfo
+      module_function
+
+      def self.names
+        CompactIndex.names(Pod.all.map(&:name))#where(:deleted => false)
```
Comment (alloy) [inline app/models/pod_info.rb]: What’s with the commented code?

File: app/models/pod_info.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+require 'compact_index'
+
+module Pod
+  module TrunkApp
+    module PodInfo
+      module_function
```
Comment (alloy) [inline app/models/pod_info.rb]: What is this for?

File: script/shard_spec_repo [hunk 8]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env ruby
+
+require 'bundler/inline'
+require 'tmpdir'
+require 'fileutils'
+
+gemfile do
+  gem 'cocoapods', '>= 1.1.0.rc.1'
```
Comment (alloy) [inline script/shard_spec_repo]: This can be bumped, although I’m sure it has no real impact.

File: script/shard_spec_repo [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      Pod::Executable.execute_command(:git, %w(push old-specs master), true)
+    end
+  end
+end
+
+Pod::Command.run(%W(repo shard master --lengths=#{prefix_lengths.join(',')}))
+
+# creating from scratch to avoid re-serializing any stale data
+Pod::Source::Metadata.from_file(master.metadata_path).tap do |metadata|
+  metadata.minimum_cocoapods_version = Pod::Version.new('1.0.0')
```
Comment (alloy) [inline script/shard_spec_repo]: Is this minimum required version correct? As in, can 1.0.0 work with sharded repos?

File: script/shard_spec_repo [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  Pod::Executable.execute_command(:heroku, %W(config:set MASTER_SOURCE_METADATA=#{yaml}), true)
+end
+
+Dir.chdir(master.repo) do
+  Pod::Executable.execute_command(:git, %W(add #{master.metadata_path}), true)
+  Pod::Executable.execute_command(:git, ['commit',
+                                         '-m', 'Update the minimum CocoaPods version to 1.0',
+                                         '-m', 'See http://blog.cocoapods.org/Master-Spec-Repo-Rate-Limiting-Post-Mortem/ for more information',
+                                         '-m', 'See http://blog.cocoapods.org/Sharding/ for the announcement of sharding',
+                                         '-m', "If you're stuck on CocoaPods < 1.0, use the `https://github.com/CocoaPods/Old-Specs.git` source from now on"],
```
Comment (alloy) [inline script/shard_spec_repo]: Nice 👌

File: script/shard_spec_repo (hunk 9 above)
Comment (segiddins) [inline script/shard_spec_repo]: 1.0.0.beta.6 could, but Orta wanted me to put in 1.0 to be on the safe side

**Comment by segiddins** [comment]
Rebased, removing the accidentally committed file

**Pull request merged**